src/lib/components/InfiniteScroll.svelte

```diff
@@ -0,0 +1,170 @@
+<script lang="ts">
+    import { onMount } from 'svelte';
+    import { PAGE_SIZE } from '$lib';
+    import { toast } from 'svelte-sonner';
+    import type { PostData, uuid } from '$lib/types';
+    import { viewport } from '$lib/util';
+
+    import Clock from 'lucide-svelte/icons/clock';
+    import ArrowUp from 'lucide-svelte/icons/arrow-up';
+    import ChevronUp from 'lucide-svelte/icons/chevron-up';
+    import ChevronDown from 'lucide-svelte/icons/chevron-down';
+
+    import Post from './Post.svelte';
+
+    import * as Select from '$lib/shadcn/components/ui/select';
+    import { Button } from '$lib/shadcn/components/ui/button';
+
+    export let data: { signed_in?: boolean } & (
+        | { source: 'home' }
+        | { source: 'channel'; channel_name: string }
+        | { source: 'private_channel'; channel_id: uuid }
+        | { source: 'user'; username: string }
+    );
+
+    const url =
+        data.source === 'home'
+            ? '/api/posts'
+            : data.source === 'channel'
+              ? `/api/c/${data.channel_name}/posts`
+              : data.source === 'private_channel'
+                ? `/api/c/private/${data.channel_id}/posts`
+                : data.source === 'user'
+                  ? `/api/u/${data.username}/posts`
+                  : 0; // unreachable
```

Comment:
Instead of having the infinite scroll component be concerned with the data source, perhaps the infinite scroll can send an event to the parent component requesting more data. That way, we also avoid using `onMount` which requires client-side fetching. We can have the first page of posts be rendered server-side creating a better experience. It'd also allow us to avoid this conditional since the server can inform the client what `url` should be.